quickwit-cluster/docs/kubernetes-discovery.md

```diff
@@ -0,0 +1,220 @@
+or network load-balancer for example).
+
+### Risk 2: Unintended cluster merges on IP reuse
+
+We want to enable users to partition their system by running several Quickwit clusters in the same Kubernetes cluster.
+Use cases are multiple:
+
+- reduce the blast radius of incidents caused by load spikes / updates / configuration changes
+- use Quickwit for several services / teams while ensuring operational isolation
+- mitigate scaling issues in the cluster membership system, if any were to come up in the future
+
+When Kubernetes pods are restarted, they are assigned a new IP and their old IP is reused by future pods.
+If an IP used by a member of `clusterA` were to be reused by a member of `clusterB`, the cluster state of
+both clusters could merge, leading to an incident.
+
+### Risk 3: Unbounded cluster state size leading to degraded performance
+
+Operating large clusters will lead to a significant churn of cluster members over time:
+
+- any configuration change or image update will result in every pod being progressively deleted and re-created,
+potentially changing their `node_id` and IP
+- both the indexer and searcher processes are good candidates for auto-scaling based on load / traffic. Auto-scaling
+would also generate a lot of member churn every day
+
+The `scuttlebutt` crate [handles this risk](https://github.com/quickwit-oss/scuttlebutt/pull/33) by garbage-collecting
+node entries if they have not been seen for 24 hours. That 24h value cloud be made configurable for a faster
+garbage collection, to support very high churn scenarios.
+
+## Illustrative Quickwit example release topology
+
+In [Helm terminology](https://helm.sh/docs/glossary/), a chart defines templates that the user will
+configure and render into a release (a set of k8s entity manifests) and apply into one k8s cluster.
+Applying several releases of the chart with different options, users can deploy several decoupled Quickwit clusters.
+Resource names are usually prefixed with the configurable release name to avoid conflicts.
+
+Defining the topology of the official Quickwit chart is out of scope for this document (especially choosing between
+deployments and statefulsets), but we will consider this potential topology (inspired from Bitnami's
+[Cassandra chart](https://github.com/bitnami/charts/tree/master/bitnami/cassandra/templates) and my experience)
+as an illustration for discussions on the proposed solution:
+
+- a `$release-indexing` deployment for an elastic pool of indexer nodes, that can be quickly auto-scaled based on volume
+- a `$release-indexing` service, to provide a DNS endpoint for the push API. This service uses a label selector to 
+  expose pods of the `$release-indexing` deployment once [they are ready](https://blog.colinbreck.com/kubernetes-liveness-and-readiness-probes-how-to-avoid-shooting-yourself-in-the-foot/)
+- a `$release-search` statefulset, for a pool of searcher nodes. Statefulset would allow prioritizing disk caching
```

Comment:
There are no disk caching on the search side at the moment. We do have RAM caching for the hotcache and column-oriented storage. Maybe we can stick to a simple deployment at the moment? Would that be more reassuring for users?

We would like to have some persistent identity however, but we just need it to be an id. 
The point there is that a node that would just be restarted would regain its node id, and take the place of its
previous instance in term of affinity with the splits. If possible, that should improve our cache efficiency.

e.g. 
Let's imagine all we need cached is cached over N searcher.
One stateful node goes down. To simplify we imagine no query is executed (but the story still holds).
The stateful node goes up but with the same node id.
All leaf request splits running on the new node will be a cache miss, but all the other nodes will have cache hits. 

With a random node id, 
All leaf request splits running on the new node will be have numsplits/N cache miss where N is the number of servers... And all splits on the new node will also be cache miss.

Reply:
It is too early to settle on this question, and a deployment would definitely work fine. I decided to use both types in my example topology to make sure the solution we discuss works with both, instead of assuming behaviours of deploy or sts.

Please note that although deployments result in random pod names, pods created from statefulsets will have a deterministic `$sts_name-$ordinal` pod name. This is one of the reasons I suggested to use the pod name as `node_id`. When the pod comes back from its update/restart, it will join back with the same `node_id` prefix, and use it's persisted disk cache to serve queries.

Reply:
My question was probably unclear.

I don't think deployment allow that,
but is it possible to have statefulsets and enjoy a sticky name, while oddly not having to configure a persisted volume.
(I think it is not possible)?

That's an easy to address but odd requirement that does not seem supported by k8s.

Reply:
Definitely, the [k8s docs]() state:

> 
> StatefulSets[ ](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#using-statefulsets)are valuable for applications that require one or more of the following.
> 
>  -   Stable, unique network identifiers.
>  -   Stable, persistent storage.
>  -   Ordered, graceful deployment and scaling.
>  -   Ordered, automated rolling updates.

I think that it's OK to define a statefulset without any `volumeClaimTemplates`, but we can try it out to confirm.